crowdsourcing/crowdsourcing_tutorial.py

```diff
@@ -19,161 +19,214 @@
+
+# %% [markdown]
+# ## Loading Crowdsourcing Dataset
 
 # %%
 import os
 
 if os.path.basename(os.getcwd()) == "snorkel-tutorials":
     os.chdir("crowdsourcing")
-from data import load_data, emotions
 
-print(f"Emotion to int mapping: {emotions}")
+# %%
+from data import load_data, answer_mapping
+
+crowd_answers, df_train, df_dev, df_test = load_data()
+Y_dev = df_dev.sentiment.values
+Y_test = df_test.sentiment.values
+
+print("Answer to int mapping:")
+for k, v in sorted(answer_mapping.items(), key=lambda kv: kv[1]):
+    print(f"{k:<50}{v}")
+
+# %% [markdown]
+# First, let's take a look at our development set to get a sense of what the tweets look like.
 
-crowd_answers, (
-    (dev_df, dev_labels),
-    (val_df, val_labels),
-    (test_df, test_labels),
-    train_df,
-) = load_data()
+# %%
+df_dev.head()
+
+# %% [markdown]
+# Now let's take a look at the crowd labels. We'll convert these into labeling functions.
+
+# %%
 crowd_answers.head()
 
 # %% [markdown]
 # ## Writing Labeling Functions
-# Each crowd worker can be thought of as a single labeling function, as each worker labels a subset of examples, and may have errors or conflicting answers with other workers / labeling functions. So we create one labeling function per worker.
+# Each crowd worker can be thought of as a single labeling function, as each worker labels a subset of examples, and may have errors or conflicting answers with other workers / labeling functions. So we create one labeling function per worker. We'll simply return the label the worker submitted for a given tweet, and abstain if they didn't submit an answer for it.
+
+# %% [markdown]
+# ### Crowd worker labeling functions
 
 # %%
-labels_by_annotator = crowd_answers[["tweet_id", "worker_id", "emotion"]].groupby(
-    ["worker_id"]
-)
+labels_by_annotator = crowd_answers.groupby("worker_id")
 worker_dicts = {}
 for worker_id in labels_by_annotator.groups:
-    worker_df = labels_by_annotator.get_group(worker_id)[["tweet_id", "emotion"]]
-    worker_dicts[worker_id] = dict(zip(worker_df.tweet_id, worker_df.emotion))
+    worker_df = labels_by_annotator.get_group(worker_id)[["answer"]]
+    v = set(worker_df.answer.tolist())
+    if len(worker_df) > 10:
+        worker_dicts[worker_id] = dict(zip(worker_df.index, worker_df.answer))
+
+print("Number of workers:", len(worker_dicts))
 
 # %%
-from snorkel.labeling.lf import LabelingFunction, labeling_function
+from snorkel.labeling.lf import LabelingFunction
 
 
-def get_worker_labeling_function(worker_id):
+def f_pos(x, worker_dict):
+    label = worker_dict.get(x.tweet_id)
+    return 1 if label == 1 else -1
+
+
+def f_neg(x, worker_dict):
+    label = worker_dict.get(x.tweet_id)
+    return 0 if label == 0 else -1
+
+
+def get_worker_labeling_function(worker_id, f):
     worker_dict = worker_dicts[worker_id]
+    name = f"worker_{worker_id}"
+    return LabelingFunction(name, f=f, resources={"worker_dict": worker_dict})
+
+
+worker_lfs_pos = [
```

Comment:
(Feel free to ignore) Until we have dependency learning, this might work worse than having a single lf, because the exclusive relationship between the positive and negative lfs for each worker is not encoded anywhere.

Reply:
Good call. Let's leave for now (for simplicity), but make a note to revisit.